#461377 [C#テキスト]10章演習問題

kurosu-being opened the pull request:
第10章 正規表現を使った高度な文字列処理

正規表現について学び、Regexクラスのなどを用いて文字列を検索する方法ついて主に学びました。

kurosu-being committed:
#461377 [C#テキスト]10章演習問題

Practice10-1-1を実装しました。

kurosu-being committed:
#461377 [C#テキスト]10章演習問題

Practice10-2-1を実装しました。

kurosu-being committed:
#461377 [C#テキスト]10章演習問題

Practice10-3-1のファイルを作成しました。

kurosu-being committed:
#461377 [C#テキスト]10章演習問題

Practice10-3-1を実装しました。
Practice10-4-1のファイルを作成しました。

kurosu-being committed:
#461377 [C#テキスト]10章演習問題

Practice10-4-1を実装しました。

kurosu-being committed:
#461377 [C#テキスト]10章演習問題

Practice10-5-1を実装しました。
Practice10-6-1のファイルを作成しました。

kurosu-being committed:
#461377 [C#テキスト]10章演習問題

Practice10-6-1を実装しました。

kurosu-being committed:
#461377 [C#テキスト]10章演習問題

Practice10-6-1において、正規表現を修正しました。

kurosu-being committed:
#461377 [C#テキスト]10章演習問題

Practice10-1-1において、メソッドにコメントを追記しました。
Practice10-2-1において、メソッドにコメントを追記しました。
Practice10-4-1において、テキストファイルをIEnumberable<T>として扱う方法に書き換えました。また、try--catchで実行中のエラーに対し例外処理を行うように変更しました。
Practice10-5-1において、開発時メモのコメントを削除しました。

kurosu-being committed:
#461377 [C#テキスト]10章演習問題

Practice10-1-1、10-2-1において、10-5-1例外処理を修正しました。
Practice10-1-1、10-2-1において、tryを修正しました。

kurosu-being committed:
#461377 [C#テキスト]10章演習問題

Practice10-4-1、において、フォーマットを修正しました。

kurosu-being committed:
#461377 [C#テキスト]10章演習問題

Practice10-1-1において、コンソールに出力するメッセージを修正しました。

kurosu-being committed:
#461377 [C#テキスト]10章演習問題

Practice10-1-1において、コンソールに出力するメッセージを修正しました。
Practice10-2-1において、コンソールに出力するメッセージを修正しました。
Practice10-4-1において、コンソールに出力するメッセージを修正しました。
Practice10-5-1において、一行ずつ区切って読み取る方法を変え、テキストファイルを一文として読み取る方法に変化させました。

kurosu-being committed:
#461377 [C#テキスト]10章演習問題

Practice10-1-1において、コンソールに出力するメッセージを修正しました。

kurosu-being committed:
#461377 [C#テキスト]10章演習問題

Practice10-3-1において、不要なコメントを削除しました。

kurosu-being committed:
#461377 [C#テキスト]10章演習問題

Practice10-2-1において、StreamReaderクラスを使って書き換えました。
Practice10-3-1において、不要なusingを削除しました。

kurosu-being committed:
#461377 [C#テキスト]10章演習問題

Practice10-5-1において、ラムダ式の中の変数名をattribute→xに変更しました。

kurosu-being committed:
#461377 [C#テキスト]10章演習問題

Practice10-5-1において、小文字への変換をメソッドに切り分けました。

k-miyahara-being commented on 章末問題10/Practice10-1-1/Program10-1-1.cs:
```diff
@@ -0,0 +1,30 @@
+        /// <param name="vInput">原文</param>
+        /// <returns>判定</returns>
+        static bool JudgePhoneNumber(string vInput) {
+            string wPattern = @"^(090|080|070)-\d{4}-\d{4}$";
```
OKです！
別解として、↓のような書き方もできます。
`@"^(0[987]0)(-\d{4}){2}$"`
090, 080, 070で「0始まり0終わり」が共通しているのでまとめた形です。
また、後半は数字4桁のブロックが2回連続するので、{2}で繰り返しを表現しています。

とはいえ、ぱっと見では黒子さんの解答のほうが理解しやすい気もします。
正規表現は、新しいケースが見つかったら都度更新していく可能性があるため、保守しやすいものになっていることが大切です。
例えばGaiaでは、設計書取込という機能があり、各発注者が用意したPDFやExcelの設計書から、単価の名称や歩掛条件などを読み込みます。
同じ意味でも、各発注者ごとに異なる言葉で表現されていることがあるため、それらを同一とみなすために正規表現が使われていたりします。

kurosu-being replied:
設計書取込機能はGaiaの大事な機能ですね。
簡潔に表現し、追加記入しやすくすることで正規表現の保守性を高めることを覚えたいと思います。
ご教示ありがとうございます！

k-miyahara-being commented on 章末問題10/Practice10-2-1/Program10-2-1.cs:
```diff
@@ -0,0 +1,57 @@
+        /// <returns>三文字以上の文字だけから成る部分文字列</returns>
+        static List<string> EnumerateSubstrings(string vText) {
+
+            var wMatches = Regex.Matches(vText, @"\b(?![0-9])\w{3,}\b");
```
問題を読み間違えていないでしょうか？「3文字以上の**数字**」じゃないかと思います。

また、「3文字以上の**文字**だけから成る部分文字列」だとしても、この正規表現では「aa111」のような先頭が数字ではないが、数字を含む文字列も通ってしまいます。

kurosu-being replied:
ご指摘ありがとうございます！
問題文を「3文字以上の数字」に修正し「aa111」のような文字列を通さないよう正規表現を修正しました。
問題文の確認を一層気を付けます。
また、テストケースについても再度様々な場面に対応できるよう勉強したいと思います。

k-miyahara-being replied:
OKです！

k-miyahara-being commented on 章末問題10/Practice10-4-1/Program10-4-1.cs:
```diff
@@ -0,0 +1,46 @@
+                Console.WriteLine("ファイルが書き換えられました");
+            } catch (Exception wEx) {
+                Console.WriteLine($"エラーが発生しました: {wEx.Message}");
+            }
```
try catchを実装したのですね！良いと思います。
ちなみに、どのような例外が起こることを想定して実装しましたか？

細かいですが、catch句の変数はコーディング規約適用外とされています。exのままでOKです

kurosu-being replied:
ご指摘ありがとうございます！
wEx→exに修正いたしました。また、正規表現に合致するものがなかった際「置換する文字列がありませんでした」と表示し終了するように修正を加えました。

try catchの例外は主にファイル操作に関連する
・アクセス権限の有無による例外
・ファイルの書き込みや削除途中にファイルが損失するなどの例外
などを想定していました。

k-miyahara-being replied:
例外の説明ありがとうございます。
学習のため、想定する例外がある場合は実際に再現してみるのも良いと思います。

> また、正規表現に合致するものがなかった際「置換する文字列がありませんでした」と表示し終了するように修正を加えました。

すでに置換したファイルをもう一度読み込んだとき、この分岐に入らないので正しい動きになっていません。
また、置き換えが発生したかどうかの結果を得るためだけにwNewLinesを全チェックするのは無駄なコストがかかってしまっている気がします。
直前のfor文の中で、同時にチェックできそうにないでしょうか？
例えば、置き換えが発生したかどうかのフラグを用意しておき、発生した時点でフラグを立てる、などです。

kurosu-being replied:
お疲れ様です。
こちらの
> すでに置換したファイルをもう一度読み込んだとき、この分岐に入らないので正しい動きになっていません。

について解消する為に、フラグを立ててforの中でチェックを行うよう修正しました。
長らくお時間空きまして恐縮ですが、ご確認いただければ幸いです🙇‍♀️

k-miyahara-being commented on 章末問題10/Practice10-4-1/Program10-4-1.cs:
```diff
@@ -0,0 +1,46 @@
+                File.WriteAllLines(wTempFilePath, wLines, Encoding.UTF8);
+
+                File.Delete(wFilePath);
+                File.Move(wTempFilePath, wFilePath);
```
①ファイルを一時保存する → ②入力したファイルを消す → ③一時保存したファイルを移動する
という流れで実装してもらったのですね。
少しキケンかなと感じたのは、②でファイルを消した後③に失敗したとき、ファイルは消されたままになってしまいます。
一時保存場所には残っているかもしれませんが、ユーザーはそのような仕様までは把握できないので「ファイルが消えた！」と感じてしまいます。
まず、問題の要件としては、ファイル自体はそのままで中身を更新するだけなのかなと思うので、Deleteを使わない方法で実装してみてください。

kurosu-being replied:
ご指摘ありがとうございます！
Deleteを使わない方法で実装する方法に修正しました。
ユーザーがファイルが消えたように見える、というのは考えられておりませんでした。データを消してしまう、というのは慎重になる必要がありますね。

k-miyahara-being commented on 章末問題10/Practice10-4-1/Program10-4-1.cs:
```diff
@@ -0,0 +1,46 @@
+
+                IEnumerable<string> wLines = File.ReadLines(wFilePath, Encoding.UTF8)
+                    .Select(line => Regex.Replace(line, @"\s*=\s*", "=").ToLower())
+                    .Select(line => Regex.Replace(line, @"version=v4.0", "version=v5.0"));
```
実行してみたところ、置き換わっていませんでした。
もう一度処理を見直してみてください。

また、テストする際はさまざまなケースを織り交ぜてみると良いです。
用意してくれたテキストファイルの中身は↓のみでしたが、これでは1ケースしかテストできません。
`version="v4.0"`

例えば、以下のようなケースを追加してテストしてみることで、実装に漏れがないか確認する必要があります。
①4.0以外の場合
②"v4.0"の"v"が大文字/小文字の場合
③"version"の"v"が大文字/小文字の場合
④"="前後に空白がある場合（前のみある、後ろのみある、前後にある）
⑤Versionの前に別の文字列がある場合
```
version="v3.0"
Version = "v4.1"
Version = "v4.0.1"
Version = "v4"
Windows version="v4.0"
MacOSVersion="v4.0"
```
例なので他にもあるかもしれません。

kurosu-being replied:
ご指摘ありがとうございます！
考えられるテストケースを書いたテキストファイルに修正し、再度チケットに貼りなおしました。
要件を正しく反映できているかどうか、テストケースを考えることもコーディングと同様に勉強していきます。

k-miyahara-being commented on 章末問題10/Practice10-5-1/Program10-5-1.cs:
```diff
@@ -0,0 +1,47 @@
+    //属性が記述されて切る場合にも対応してください。属性の中には'<','>'は含まれていないものとします。
+    class Program {
+        static void Main(string[] args) {
+            Console.WriteLine("HTMLファイルを読み込んでください。サンプルのHTMLファイルはチケットに貼ってあります。");
+            var wFilePath = Console.ReadLine();
```
チケット添付のファイルですが、HTMLファイルではなくテキストファイルとなっています。
HTMLファイルにまだ出会っていないかもしれませんが、拡張子が「.html」のものに置き換えてみてください。

kurosu-being replied:
ご指摘ありがとうございます！
HTMLファイルに変更し、修正いたしました。

k-miyahara-being commented on 章末問題10/Practice10-5-1/Program10-5-1.cs:
```diff
@@ -0,0 +1,47 @@
+        /// <param name="vHtmlContent">HTMLの原文</param>
+        /// <returns> 属性のタグを小文字にしたテキスト</returns>
+        static string ConvertToLowerHTML(string vHtmlContent) {
+            string wModifiedContent = Regex.Replace(vHtmlContent, @"<[^>]+>", match => {
+                string wTag = match.Value.ToLower();
+                wTag = Regex.Replace(wTag, @"(\w+)(?=\s*=)", x => x.Value.ToLower());
+                return wTag;
+            });
```
↓のように、タグ以外の部分(myBox, Myid)の大文字も小文字に変換されてしまいます。
`<DIV class="myBox" id="Myid">` → `<div class="mybox" id="myid">`

DIVのみ小文字に変換されるように修正してみてください。

kurosu-being replied:
ご指摘ありがとうございます！
DIVのようなタグのみ小文字に変換されるように修正を行いました。

kurosu-being committed:
#461377 [C#テキスト]10章演習問題

Practice10-2-1において、問題文を修正し、3文字以上の文字だけからなる部分文字列→3文字以上の数字だけからなる部分文字列に変更いたしました。
Practice10-4-1において、wEx→exに修正いたしました。

kurosu-being committed:
#461377 [C#テキスト]10章演習問題

Practice10-2-1において、正規表現を修正しました。

kurosu-being committed:
#461377 [C#テキスト]10章演習問題

Practice10-4-1において、正規表現委合致するものがなかった際「置換する文字列がありませんでした」と表示し終了するように修正を加えました。

kurosu-being committed:
#461377 [C#テキスト]10章演習問題

Practice10-4-1において、LINQの表現を変更しました。

kurosu-being committed:
#461377 [C#テキスト]10章演習問題

Practice10-5-1において、タグのみを小文字変換するように正規表現を修正しました。

k-miyahara-being commented on 章末問題10/Practice10-2-1/Program10-2-1.cs:
```diff
@@ -0,0 +1,59 @@
+                List<int> wSubstrings = EnumerateSubstrings(wText);
+
+                if (wSubstrings.Count > 0) {
+                    Console.WriteLine("3文字以上の部分文字列:");
```
リストの要素が存在するかどうかは、引数なしのAnyメソッドを使うとよいです。

kurosu-being replied:
ご指摘ありがとうございます。
引数なしのAnyメソッドを用いて修正しました。

kurosu-being committed:
#461377 [C#テキスト]10章演習問題

Practice10-2-1において、リストの要素が存在するかどうかを引数なしのAnyメソッドで調べるように修正しました。

kurosu-being committed:
#461377 [C#テキスト]10章演習問題

Practice10-4-1において、例外の処理を明示的に行いました。
Practice10-4-1において、置き換えが行われたかどうかをフラグを立てることで判定するように変更しました。

kurosu-being committed:
#461377 [C#テキスト]10章演習問題

Practice10-4-1において、例外のメッセージにユーザーへの提案を追加しました。

kurosu-being committed:
#461377 [C#テキスト]10章演習問題

Practice10-4-1において、wNewLinesを削除し、フラグによって書き込みを行うかの条件分岐に入るよう修正しました。

kurosu-being commented:
@k-miyahara-being 
https://github.com/kurosu-being/TextC/pull/11#discussion_r1429540117
（直接返信が出来なかったので、引用にて失礼します）

ご指摘ありがとうございました。
forの中でwWasReplacedのフラグを判定するように変更いたしました。
自身でも確認いたしましたが、再度確認いただけると幸いです。
